README.md

```diff
@@ -99,36 +99,77 @@ book].
 
 #### Android
 
-This library makes use of the [ndk-rs](https://github.com/rust-windowing/android-ndk-rs) crates, refer to that repo for more documentation.
+The Android backend builds on (and exposes types from) the [`ndk`](https://docs.rs/ndk/0.7.0/ndk/) crate.
 
-The `ndk-glue` version needs to match the version used by `winit`. Otherwise, the application will not start correctly as `ndk-glue`'s internal `NativeActivity` static is not the same due to version mismatch.
+Native Android applications need some form of "glue" crate that is responsible for defining the main entry point for your Rust application as well as tracking various life-cycle events and synchronizing with the main JVM thread.
 
-`winit` compatibility table with `ndk-glue`:
+Winit uses the [android-activity](https://github.com/rib/android-activity) as a glue crate (prior to `0.28` it used [ndk-glue](https://github.com/rust-windowing/android-ndk-rs/tree/master/ndk-glue).
 
-| winit |       ndk-glue       |
-| :---: | :------------------: |
-| 0.24  | `ndk-glue = "0.2.0"` |
-| 0.25  | `ndk-glue = "0.3.0"` |
-| 0.26  | `ndk-glue = "0.5.0"` |
-| 0.27  | `ndk-glue = "0.7.0"` |
+The version of the glue crate that your application depends on _must_ match the version that Winit depends on because the glue crate is responsible for your application's main entrypoint. If Cargo resolves multiple versions they will clash.
```

Comment:
Wouldn't it be possible to just re-export the interesting parts from `android-activity` including the `native-activity`/`game-activity` flags? This would remove the requirement for users to have a compatible crate version.

Reply:
NB: Android applications are always JVM based and start with a JVM `Activity` which is subclassed to be able to create "native" applications by loading a shared library via `dlopen`/`LoadLibrary`.

The shared library needs to export a standard symbol that the Activity class can find as the entry point for running the native C/C++/Rust code for the application.

The 'glue' layer/crate (e..g `ndk_glue` or `android-activity`) is what provides that symbol which is the entry point for the application, and there can technically only be a single instance of that symbol.

The ABI / details for exactly how that symbol should be defined also depends on the implementation details of the Activity subclass (such as `NativeActivity` vs `GameActivity`) which conflict and mean it's not possible to support both Activity base classes in a single share library.

The state that the glue layer works with is all bootstrapped based on state passed from the JVM Activity into that entrypoint which also means that if you were to somehow get an application to link multiple instances of the glue layer and avoid a symbol clash for the entrypoint then only one of them actually going to passed the state it needs to operate.

Conceptually this is a bit different to shared libraries where multiple version can be completely orthogonal.


Differences in `NativeActivity` and `GameActivity` lead to quite significant differences in the Rust/native glue layer, especially for input handling which currently also leads to using (exposed in the public API) completely different types for input events (for motion and key events) that aren't runtime compatible.

It's sort of comparable to how winit backends aren't designed to be runtime selectable, which is reasonable from the pov that you only need one at a time. For `android-activity` we only need to support one Activity class at a time so it only strives for API compatibility across different Activity base classes - not for dynamic, runtime API + ABI compatibility.

Hope that clarifies a bit the unique challenge with glue crates here.

Reply:
> and there can technically only be a single instance of that symbol.

Such a situation should lead to a linker error?

Fwiw the text quoted here is what I set out to fix with aforementioned work on `ndk-glue`, to deprecate the `ndk-context` crate again (which is holding untyped `static` global pointers), and rely on runtime passing of all the various _typed_ NDK instances/pointers. Any mismatch in crate versions would result in a compile error (unless the user somehow manages to combine them _and_ forget to initialize `winit` with said instance through an extension `trait`).

Reply:
@rib thanks for the detailed write up but I believe I failed to get my point across or I still miss some part:
My proposal would be that the user doesn't have to basically know anything about the underlying `android-activity` crate acting as backend for the android implementation.
```toml
[dependencies]
winit = { version = "0.x", features = ["android-native-activity"] } # internally enables the `android-activity/native-activity` feature
```

instead of  
```toml
[dependencies]
winit = "0.x"
android-activity = { version = "0.y", features = ["native-activity"] }
```

Under `platform/android` public api (ie `AndroidApp`, etc) would be re-exported if necessary. I tried it locally with `NativeActivity` and this seems to work in a quick prototype given the entrypoints are specified in the `android-activity` crate with the feature flags passed trough by `winit`

Reply:
My initial instinct here is that this feels like a layering violation - though I can see the appeal of having just one thing specifying a version for `android-activity` (could reduce churn in case of API bumps)

My main concern is that the glue crate fundamentally defines the ABI for the entrypoint for the application and as you noted there is the `AndroidApp` type that's passed to the application's `android_main`. I think it would seem quite odd to me that I'm getting the definition for a type passed to my application `main` from a middleware/framework crate.

The responsibilities to decide what `Activity` class is used and the ABI for the application entry point seem like they should lay directly with the application crate and everything is downstream from that.

Please also keep in mind that android-activity is intended to be usable for native Rust apps that don't necessarily use Winit, For example I'm poking at using it with OpenXR, not using Winit in that case. (So my general mindset is to associate android-activity with the orthogonal definition of the application entrypoint before considering that framework the app is built on)

I'm currently not quite sure how this would work for something like an egui-eframe or Bevy application where Winit is mostly abstracted away from the application (except there is still a requirement to pass that `AndroidApp` in to eframe for it to construct its Winit event loop).

In general if there are apps using Winit behind an abstraction then you'd still have the original requirement for the application crate to have a direct dependency on the android-activity crate - or else those abstractions would also need to re-export `AndroidApp` in a similar way.

I guess I'm not quite sure either way about this idea atm. I can see the benefit that some apps they could just defer to winit to specify the android-activity version but I can't help but find it a bit odd that a key type for the main entry point is coming from a higher-level middleware crate.

Reply:
> > and there can technically only be a single instance of that symbol.
> 
> Such a situation should lead to a linker error?

It depends:

In the case of `ndk-glue` it assumes you're using a proc macro that actually makes the application crate define the symbol that `NativeActivity` looks for, so there won't be a linker clash and things will only go wrong at runtime because only the version of `ndk-glue` that's linked to the application will get initialized by that entrypoint. If a crate like winit has resolved a different version then when it tries to query some state it will be interacting with an uninitialized version of `ndk-glue`.

In the case of `android-activity` there will generally be a compile-time error before linking because the feature that determines which `Activity` backend to use is unlikely to be specified for additional versions and it's a compile-time error if no Activity backend has been chosen. If you did somehow specify a backend for each version then yep you should still hit a linker error while compiling.

Reply:
Just to follow up again on this idea of yours @msiglreith... Since this seems like an orthogonal change we could make later; can we maybe agree to re-visit this as a separate issue/PR?

I agree that it could be an ergonomics improvement in some cases to re-export `AndroidApp` via the Winit crate if that means application crates don't need to explicitly declare their own dependence on `android-activity`.

More generally I think the information about the `android-activity` version needing to match what Winit expects is still valid (similar to how it had to match with ndk-glue) and it seems worthwhile to document this. This idea to re-export `AndroidApp` should hopefully just make it less likely that developers will get caught out by this requirement.

Reply:
I've just updated this branch to include this suggestion.

The `android-activity` API is now re-exported under `winit::platform::android::activity::*` and there are now "android-native-activity" and "android-game-activity" features that set the corresponding features in `android-activity`.

This way it's possible for application crates to avoid explicitly adding an `android-activity` dependency which could lead to a version conflict any time winit bumps its `android-activity` version.

The README reflects this too.